x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -281,4 +284,36 @@ public void optimizedPhysicalPlan(LogicalPlan logicalPlan, ActionListener<Physic
             return plan;
         }));
     }
+
+    public static InvalidMappedField specificValidity(String fieldName, Map<String, FieldCapabilities> types) {
+        boolean hasUnmapped = types.containsKey(UNMAPPED);
+        boolean hasTypeConflicts = types.size() > (hasUnmapped ? 2 : 1);
+        String metricConflictsTypeName = null;
+        boolean hasMetricConflicts = false;
+
+        if (hasTypeConflicts == false) {
+            for (Map.Entry<String, FieldCapabilities> type : types.entrySet()) {
+                if (UNMAPPED.equals(type.getKey())) {
+                    continue;
+                }
+                if (type.getValue().metricConflictsIndices() != null && type.getValue().metricConflictsIndices().length > 0) {
+                    hasMetricConflicts = true;
```

Comment:
i'd find it slightly easier to read if the `InvalidMappedField` would be built here.